voxel_grid/prototype/mvp_v2.hpp

```diff
@@ -0,0 +1,243 @@
+
+
+#include <Eigen/Dense>
+#include <boost/optional.hpp>
+#include <iostream>
+#include <unordered_map>
+#include <vector>
+
+using Vec3f = Eigen::Vector3f;
+
+class PointXYZ {
+ public:
+  PointXYZ(const float x = 0, const float y = 0, const float z = 0) {
+    p << x, y, z;
+  }
+  PointXYZ(const Vec3f &p) : p(p) {}
+  Vec3f p;
+};
+
+bool isXYZFinite(const PointXYZ &pt) { return true; }
+
+template <typename PointT>
+class PointCloudT {
+ public:
+  PointCloudT() {}
+
+  void push_back(const PointT &p) { points.push_back(p); }
+  void reserve(const size_t s) { (void)s; }
+  auto begin() { return points.begin(); }
+  auto end() { return points.end(); }
+
+  // Defined in pcl::PointCloud
+  using PointType = PointT;
+
+  bool is_dense;
+  std::vector<PointT> points;
+};
+
+template <typename PointT>
+class Filter {
+  using PointCloud = PointCloudT<PointXYZ>;
+
+ public:
+  Filter() {}
+  void filter(PointCloud &output) { applyFilter(output); }
+
+ protected:
+  virtual void applyFilter(PointCloud &pc) = 0;
+  PointCloud input_;
+};
+
+// ================== transform_filter.hpp ==================
+
+#define GET_POINT_TYPE(GridStructT) typename GridStructT::PointCloud::PointType
+
+template <typename GridStruct, typename PointT = GET_POINT_TYPE(GridStruct)>
+class TransformFilter : public Filter<PointT> {
```

Comment:
As mentioned by me elsewhere, we should create a uniform interface for template types for FunctionFilter and TransformFilter (both should have either PointT  first or second). Smthg to discuss, but not here (just mentioning so I don't forget it)